oak_functions_service/src/lookup_htbl.rs

```diff
@@ -0,0 +1,575 @@
+// limitations under the License.
+//
+
+// This is a key/value hash table optimized for large numbers of small k/v pairs.  We assume we
+// bust out of on-chip cache, so speed is limited by cache misses, not the actual operations such
+// as hashing and bit manipulation.
+//
+// The k/v pairs are loaded before any are read, and once they are loaded cannot be modified.  We
+// take advantage of these restrictions to build a hash table with far less memory overhead per k/v
+// pair (21 bytes vs ~128 bytes for hashbrown<Bytes, Bytes>).  Speed for k/v lookups is ~90ns on
+// manual tests.
+
+use alloc::{boxed::Box, vec, vec::Vec};
+use core::mem;
+
+use bytes::Bytes;
+
+// To save memory, we use 5 byte array index values instead of 8 bytes.  This saves 12 bytes per
+// k/v pair.  It is referred to as u40 below.
+const INDEX_SIZE: usize = 5;
+// Size of an entry in the hash table.
+const ENTRY_SIZE: usize = INDEX_SIZE + 1;
+
+// The hash table consists of 2 vectors.  The first is a table of "entries", which is a packed
+// array of entries with the following layout:
+//
+//    hash_byte: u8,
+//    data_index: u40,  // A 5-byte index, so we support up to 1TiB of data.
+//
+// This is a "swiss" table meaning that instead of the usual "buckets" of entries, the entries are
+// collapsed into the hash table itself.  By adding a 1-byte hash, we can eliminate over 99.5% of
+// key comparisons to the wrong keys.  This is important for speed because comparing keys causes a
+// page fault usually, while comparing thie 1-byte hash is usually an L1 cache hit.  The number of
+// cache misses per lookup should be close to 2 for values found in the table, and 1 for values not
+// in the table.
+//
+// The storage for k/v pairs is chunked into 256 4GiB vectors:
+//
+//     data_chunks: Vec<Box([u8; N])>,  // N = 2^32
+//     used_chunks: usize,
+//
+// data_chunks starts empty, and as data is added to the table, we allocat another chunk.
+// The format for each k/v pair of data in each chunk is:
+//
+//     key_len: <compressed u30>
+//     key: [u8]
+//     value_len: <compressed u30>
+//     value: [u8]
+//
+// Another memory savings comes from using a swiss table with exactly a 60% load factor.  Swiss
+// tables require a significant portion of the table be empty, and normally as the table grows,
+// they dynamically double the size.  At least when using the alloc crate, this resize causes a new
+// table to be allocated at 2X the size, data is copied from the old table to the new, and the old
+// table is freed.  This results in a 3X expansion in memory usage vs the prior array size.  In
+// contrast, this scheme has only 40% of the table empty and is never resized.
+//
+// This reduce scheme currently limits the number of entries to 2^32.
+
+const CHUNK_BITS: usize = 21;
+const CHUNK_SIZE: usize = 1 << CHUNK_BITS;
+const CHUNK_MASK: usize = CHUNK_SIZE - 1;
+
+#[derive(Default)]
+pub struct LookupHtbl {
```

Comment:
Add some documentation for the fields (and the enum, below).

Reply:
Done.